ambry-store/src/main/java/com/github/ambry/store/PersistentIndex.java

```diff
@@ -1290,7 +1279,10 @@ FindInfo findEntriesSince(FindToken token, long maxTotalSizeOfEntries) throws St
           // Get entries starting from the first key in this offset.
           Map.Entry<Offset, IndexSegment> entry = indexSegments.floorEntry(offsetToStart);
           StoreFindToken newToken;
-          if (entry != null && !entry.getKey().equals(indexSegments.lastKey())) {
+          //if the token is point to the last index entry of last index segment and journal cleaned up after last log segment
+          //get auto closed, we should transform the journal based token into index based token.
+          if (entry != null && !entry.getKey().equals(indexSegments.lastKey()) || onlyContainsHeaderInLastLogSegment(
+              offsetToStart, indexSegments)) {
```

Comment:
I think this deserves more comments, like

The reasons why we are here could be 
1. StoreFindToken is Uninitialized, and we have more index segments than journal can cover
2. StoreFindToken is JournalBased, but the journal somehow moved to much ahead, the offset in the FindToken it no longer within journal.
    2.1 There might just be lots of new entries coming to journal so the journal moves ahead
    2.2 The journal might be cleared and new log segment is created for compaction (auto-close last segment)

These would be three reasons why we are here.
For reason 1, entry equals to first index segment and doesn't equals to last segment (we have more index segments  than journal can cover)
For reason 2.1. entry equals to some index segment in the middle, but not last segment
For reason 2.2 entry might equal to last segment is the new log segment doesn't have any new record.

That's why we are adding the new check  here.

And I have no idea how we can reach line 1297.

Reply:
If token reached to the last entry of the last index segment, and the last log segment has not been auto closed, which means token reached to the end of current log segment, at this time, we should return the original journal based token instead of transform it into index based.(line 1297)

Reply:
Updated the comments. I think when at beginning log segment is empty and index segment is empty, the start offset would be 18 and entry will be null since journal is null, the Uninitialized should be return directly which will reach line 1297.